cmd/derper/derpprobe/derpprobe.go

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// The derpprobe binary probes derpers.
+package main // import "tailscale.com/cmd/derper/derpprobe"
+
+import (
+	"bytes"
+	"context"
+	crand "crypto/rand"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"html"
+	"io"
+	"log"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"tailscale.com/derp"
+	"tailscale.com/derp/derphttp"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/key"
+)
+
+var (
+	derpMapURL = flag.String("derp-map", "https://login.tailscale.com/derpmap/default", "URL to DERP map")
+	listen     = flag.String("listen", ":8030", "HTTP listen address")
+)
+
+var (
+	mu            sync.Mutex
+	state         = map[nodePair]pairStatus{}
+	lastDERPMap   *tailcfg.DERPMap
+	lastDERPMapAt time.Time
+)
+
+func main() {
+	flag.Parse()
+	go probeLoop()
+	log.Fatal(http.ListenAndServe(*listen, http.HandlerFunc(serve)))
+}
+
+func serve(w http.ResponseWriter, r *http.Request) {
+	mu.Lock()
+	defer mu.Unlock()
+	if lastDERPMap == nil {
+		http.Error(w, "no DERP map", 500)
+		return
+	}
+	now := time.Now()
+	var bad []string
+	var good []string
+	if age := now.Sub(lastDERPMapAt); age > time.Minute {
+		bad = append(bad, fmt.Sprintf("DERPMap hasn't been successfully refreshed in %v", age.Round(time.Second)))
+	}
+	for _, reg := range sortedRegions(lastDERPMap) {
+		for _, from := range reg.Nodes {
+			for _, to := range reg.Nodes {
+				pair := nodePair{from.Name, to.Name}
+				st, ok := state[pair]
+				age := now.Sub(st.at).Round(time.Second)
+				switch {
+				case !ok:
+					bad = append(bad, fmt.Sprintf("no state for %v", pair))
+				case st.err != nil:
+					bad = append(bad, fmt.Sprintf("%v: %v", pair, st.err))
+				case age > 90*time.Second:
+					bad = append(bad, fmt.Sprintf("%v: update is %v old", pair, age))
+				default:
+					good = append(good, fmt.Sprintf("%v: %v, %v ago", pair, st.latency.Round(time.Millisecond), age))
+				}
+			}
+		}
+	}
+	summary := "All good"
+	if len(bad) > 0 {
+		w.WriteHeader(500)
+		summary = fmt.Sprintf("%d problems", len(bad))
+	}
+	io.WriteString(w, "<html><head><style>.bad { font-weight: bold; color: #700; }</style></head>\n")
+	fmt.Fprintf(w, "<body><h1>derp probe</h1>\n%s:<ul>", summary)
+	for _, s := range bad {
+		fmt.Fprintf(w, "<li class=bad>%s</li>\n", html.EscapeString(pretty(s)))
+	}
+	for _, s := range good {
+		fmt.Fprintf(w, "<li>%s</li>\n", html.EscapeString(pretty(s)))
+	}
+}
```

Comment:
`fmt.Fprintf(w, "</ul></body></html>")`

no good reason other than to stop my eye twitching

Reply:
embrace HTML5, man. so many implicit closes :)

but done